llvm/lib/CAS/MappedFileRegionBumpPtr.cpp

```diff
@@ -146,66 +157,79 @@ Expected<MappedFileRegionBumpPtr> MappedFileRegionBumpPtr::create(
   if (!FileSize)
     return createFileError(Result.Path, FileSize.getError());
 
+  // If the size is smaller than the capacity, we need to initialize the file.
+  // It maybe empty, or may have been shrunk during a previous close.
   if (FileSize->Size < Capacity) {
     // Lock the file exclusively so only one process will do the initialization.
-    if (Error E = InitLock.unlock())
-      return std::move(E);
     if (Error E = InitLock.lock(sys::fs::LockKind::Exclusive))
       return std::move(E);
     // Retrieve the current size now that we have exclusive access.
     FileSize = FileSizeInfo::get(File);
     if (!FileSize)
-        return createFileError(Result.Path, FileSize.getError());
+      return createFileError(Result.Path, FileSize.getError());
   }
 
-  // At this point either the file is still under-sized, or we have the size for
-  // the completely initialized file.
-
-  if (FileSize->Size < Capacity) {
-    // We are initializing the file; it may be empty, or may have been shrunk
-    // during a previous close.
-    // FIXME: Detect a case where someone opened it with a smaller capacity.
+  uint64_t MappingSize = FileSize->Size;
+  // If the size is still smaller than the minimal required size, we need to
+  // resize the file to the capacity.
+  if (FileSize->Size < MinCapacity) {
     assert(InitLock.Locked == sys::fs::LockKind::Exclusive);
     if (std::error_code EC = sys::fs::resize_file_sparse(FD, Capacity))
       return createFileError(Result.Path, EC);
-
     if (Result.Logger)
       Result.Logger->log_MappedFileRegionBumpPtr_resizeFile(
           Result.Path, FileSize->Size, Capacity);
-  } else {
-    // Someone else initialized it.
-    Capacity = FileSize->Size;
+    MappingSize = Capacity;
   }
 
   // Create the mapped region.
   {
     std::error_code EC;
     sys::fs::mapped_file_region Map(
-        File, sys::fs::mapped_file_region::readwrite, Capacity, 0, EC);
+        File, sys::fs::mapped_file_region::readwrite, MappingSize, 0, EC);
     if (EC)
       return createFileError(Result.Path, EC);
     Result.Region = std::move(Map);
   }
 
-  if (FileSize->Size == 0) {
-    assert(InitLock.Locked == sys::fs::LockKind::Exclusive);
-    // We are creating a new file; run the constructor.
+  if (FileSize->Size < MinCapacity) {
+    // If we need to fully initialize the file, call NewFileConstructor.
     if (Error E = NewFileConstructor(Result))
-      return std::move(E);
-  } else {
-    Result.initializeBumpPtr(BumpPtrOffset);
+      return E;
```

Comment:
`std::move(E)` is required by some supported compilers when the return type is `Expected` rather than `Error`.